utils/s3/client.cc

```diff
@@ -0,0 +1,123 @@
+    co_await util::skip_entire_stream(in);
+}
+
+client::client(socket_address addr, private_tag)
+        : _addr(std::move(addr))
+        , _host(to_sstring(_addr))
+        , _http(_addr)
+{
+}
+
+shared_ptr<client> client::make(socket_address addr) {
+    return seastar::make_shared<client>(std::move(addr), private_tag{});
+}
+
+future<uint64_t> client::get_object_size(sstring object_name) {
+    s3l.trace("HEAD {}", object_name);
+    auto req = http::request::make("HEAD", _host, object_name);
+    uint64_t len = 0;
+    co_await _http.make_request(std::move(req), [&len] (const http::reply& rep, input_stream<char>&& in_) mutable -> future<> {
+        len = rep.content_length;
+        return make_ready_future<>(); // it's HEAD with no body
+    });
+    co_return len;
+}
+
+future<temporary_buffer<char>> client::get_object_contiguous(sstring object_name, std::optional<range> range) {
+    auto req = http::request::make("GET", _host, object_name);
+    http::reply::status_type expected = http::reply::status_type::ok;
+    if (range) {
+        auto range_header = format("bytes={}-{}", range->off, range->off + range->len - 1);
+        s3l.trace("GET {} contiguous range='{}'", object_name, range_header);
+        req._headers["Range"] = std::move(range_header);
+        expected = http::reply::status_type::partial_content;
+    } else {
+        s3l.trace("GET {} contiguous", object_name);
+    }
+
+    size_t off = 0;
+    std::optional<temporary_buffer<char>> ret;
+    co_await _http.make_request(std::move(req), [&off, &ret, &object_name] (const http::reply& rep, input_stream<char>&& in_) mutable -> future<> {
+        auto in = std::move(in_);
+        ret = temporary_buffer<char>(rep.content_length);
+        s3l.trace("Consume {} bytes for {}", ret->size(), object_name);
+        co_await in.consume([&off, &ret] (temporary_buffer<char> buf) mutable {
+            if (buf.empty()) {
+                return make_ready_future<consumption_result<char>>(stop_consuming(std::move(buf)));
+            }
+
+            size_t to_copy = std::min(ret->size() - off, buf.size());
+            if (to_copy > 0) {
+                std::copy_n(buf.get(), to_copy, ret->get_write() + off);
+                off += to_copy;
+            }
+            return make_ready_future<consumption_result<char>>(continue_consuming());
+        });
+    }, expected);
+    ret->trim(off);
+    s3l.trace("Consumed {} bytes of {}", off, object_name);
+    co_return std::move(*ret);
+}
+
+future<> client::put_object(sstring object_name, temporary_buffer<char> buf) {
+    s3l.trace("PUT {}", object_name);
+    auto req = http::request::make("PUT", _host, object_name);
+    auto len = buf.size();
+    req.write_body("bin", len, [buf = std::move(buf)] (output_stream<char>&& out_) mutable -> future<> {
+        auto out = std::move(out_);
+        co_await out.write(buf.get(), buf.size());
+        co_await out.flush();
+        co_await out.close();
```

Comment:
The close() needs protection in case of an exception, no?

Reply:
Hm... I thought close() shouldn't throw not to produce more scars to Benny. What am I missing?

Reply:
I meant that close() should be called even if one of the previous calls throws. Totally against allowing close() to throw, we need to preserve @bhalevy sanity.